bestie/server/testresult.go

```diff
@@ -0,0 +1,220 @@
+}
+
+type metricTestResult struct {
+	table   bigQueryTable
+	metrics []testMetric
+}
+
+// Pertinent fields to uniquely identify a Bazel event stream.
+type bazelStream struct {
+	buildId       string
+	invocationId  string
+	testName      string
+	run           string
+	invocationSha string // derived
+}
+
+// Derive a unique invocation SHA value.
+func deriveInvocationSha(items []string) string {
+	// The SHA value allows combining one or more items
+	// (e.g. invocationId, buildId, run) to uniquely identify
+	// a bazel test stream without impacting the database schema.
+	hash := sha256.Sum256([]byte(strings.Join(items[:], ".")))
+	return hex.EncodeToString(hash[:])
+}
+
+// Store fields that help identify this Bazel stream.
+func identifyStream(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) bazelStream {
+	// Extract the stream identifier fields of interest.
+	stream := bazelStream{
+		buildId:      streamId.GetBuildId(),
+		invocationId: streamId.GetInvocationId(),
+		run:          strconv.Itoa(int(bazelBuildEvent.GetId().GetTestResult().GetRun())),
+		testName:     bazelBuildEvent.GetId().GetTestResult().GetLabel(),
+	}
+	// Calculate a SHA256 hash using the following fields to uniquely identify this stream.
+	stream.invocationSha = deriveInvocationSha([]string{stream.invocationId, stream.buildId, stream.run})
+	return stream
+}
+
+// Handle metrics extraction from the TestResult event.
+func handleTestResultEvent(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) error {
+	stream := identifyStream(bazelBuildEvent, streamId)
+	m := bazelBuildEvent.GetTestResult()
+	if m == nil {
+		return fmt.Errorf("Error extracting TestResult data from event message")
+	}
+
+	fmt.Printf("TestResult for %s: %s\n", stream.testName, m.GetStatus())
+	fmt.Printf("  run: %s\n", stream.run)
+	fmt.Printf("  buildId: %s\n", stream.buildId)
+	fmt.Printf("  invocationId: %s\n", stream.invocationId)
+	fmt.Printf("  invocationSha: %s\n\n", stream.invocationSha)
+
+	outputFiles := m.GetTestActionOutput()
+	var ofname, ofuri string
+	found := false
+	for _, of := range outputFiles {
+		ofname = of.GetName()
+		ofuri = of.GetUri()
+		if strings.HasSuffix(ofname, "outputs.zip") {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	// Strip off any file:// prefix from the URI to access the local file system path.
+	filePrefix := "file://"
+	if strings.HasPrefix(ofuri, filePrefix) {
+		ofuri = ofuri[len(filePrefix):]
+	}
+
+	// Process test metrics output file(s).
+	// Each output file contains a single (potentially large) protobuf message.
+	if err := extractZippedFiles(stream, ofuri); err != nil {
+		return fmt.Errorf("Error processing %s file: %s", ofname, err)
+	}
+
+	return nil
+}
+
+// Look for and return a []byte slice with its contents.
+func extractZippedFiles(stream bazelStream, zipFile string) error {
+	metricsFileRE := regexp.MustCompile(`(^test_metrics.pb$|^test_metrics[_-]+[[:word:]-]*[[:alnum:]]\.pb$)`)
```

Comment:
We should pick a convention that's easier to check for (e.g. "file must end in `.metrics.pb`") to simplify our regex checking

Reply:
OK, I'll change this in the current PR before merging.